earthaccess-draft (#175)
nsidc/earthaccess: README.md

    <img src="https://readthedocs.org/projects/earthdata/badge/?version=latest&style=plastic" alt="Documentation link">
</a>

</p>

## Overview

> The real power of open science in the age of cloud computing is only unleashed to its full potential if we have easy-to-use workflows that facilitate working with data in an inclusive, efficient and reproducible way. Unfortunately —as it stands today— scientists are facing a steep learning curve for unintended complex systems and end up spending more time on the technicalities of the cloud and NASA APIs rather than focusing on their important science.
> 
> During several workshops organized by [NASA Openscapes](https://nasa-openscapes.github.io/events.html) the need to provide easy-to-use tools to our users became evident. We shouldn’t need to be cloud infrastructure engineers to easily work with the data if it’s hosted on a S3 bucket. This was the main reason behind *earthaccess*, a Python library that aims to simplify our workflows to address some of the pain points of programmatic access to NASA datasets within the PyData ecosystem.
> 
> With *earthaccess* we can login with NASA, search and download data with a few lines of code and even more relevant, our code will work the same way if we are running it in the cloud or from our laptop. ***earthaccess*** handles authentication with [NASA's Earthdata Login (EDL)](https://urs.earthdata.nasa.gov), search using NASA's [CMR](https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html) and access through [`fsspec`](https://github.com/fsspec/filesystem_spec). If all of this sounds like gibberish... no worries we are here to help! Please ask us anything on [project board!](https://github.com/nsidc/earthdata/discussions).
> 
> The only requirement to use this library is to open a free account with NASA [EDL](https://urs.earthdata.nasa.gov).

Reviewer: ```suggestion
NASA earth science data is migrating to the cloud and users shouldn’t need to be cloud infrastructure engineers to work with data in the cloud. Discovering and accessing data should be easy, so more time can be spent working with the data. During several workshops organized by [NASA Openscapes](https://nasa-openscapes.github.io/events.html) it became evident easy-to-use tools are needed to facilitate working with data in an inclusive, efficient and reproducible way. These are the motivations behind ***earthaccess***, a Python library that aims to simplify our workflows to address some of the pain points of programmatic access to NASA datasets within the PyData ecosystem. 

***earthaccess*** handles authentication with [NASA's Earthdata Login (EDL)](https://urs.earthdata.nasa.gov), search using NASA's [CMR](https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html) and access through [`fsspec`](https://github.com/fsspec/filesystem_spec). If all of this sounds like gibberish... no worries we are here to help! Please ask us anything on [project board!](https://github.com/nsidc/earthdata/discussions).
```

> The only requirement to use this library is to open a free account with NASA [EDL](https://urs.earthdata.nasa.gov).

Reviewer: This line is already mentioned above so I think it can be removed from here.

<a href="https://urs.earthdata.nasa.gov"><img src="https://auth.ops.maap-project.org/cas/images/urs-logo.png" /></a>

<img src="https://user-images.githubusercontent.com/717735/199083271-79ce5113-b680-4fe5-a1bf-6c927bf02c75.jpg" />

## Installing earthaccess

Install the latest release using conda

```bash
conda install -c conda-forge earthaccess
```

Using Pip

```bash
pip install earthaccess
```

Try it in your browser without installing anything! [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/nsidc/earthdata/main)

## Do where do we start?

Data Pathfinders guide users through the process of selecting application-specific datasets and learning how to use them through intuitive tools, facilitating equal and open access to the breadth of NASA Earth science data. The pathfinders are intended to familiarize users with and provide direct links to the applicable, commonly-used datasets across NASA’s Earth science data collections. After getting started here, there are numerous NASA resources that can help develop one’s skills further.

[https://www.earthdata.nasa.gov/learn/pathfinders](

> https://www.earthdata.nasa.gov/learn/pathfinders)

Reviewer: I suggest removing this section, based on the 'Art of the readme' where they recommend making it as short as it needs to be. I think this section makes it a bit too long and might confuse users, I think it is best just to focus on earthaccess instead of linking to other datasets and tools.

## Authentication

> After we get a new account with EDL we can authenticate ourselves using 3 different methods:

Reviewer: ```suggestion
Once you have an EDL account, you can authenticate using one of the following three methods:
```

1. Using a `.netrc` file

>     * We can use *earthaccess* to read our credentials fom a netrc file

Reviewer: ```suggestion
    * Can use *earthaccess* to read your EDL credentials (username and password) from a `.netrc` file
```

> 2. Reading our credentials from environment variables
>     * if available we can use environment variables **EDL_USERNAME** and **EDL_PASSWORD**
> 3. Interactively entering our credentials
>     * We can ask for these credentials and even persist them to a `.netrc` file

Reviewer: ```suggestion
2. Reading your EDL credentials from environment variables
    * if available you can use environment variables **EDL_USERNAME** and **EDL_PASSWORD**
3. Interactively entering your EDL credentials
    * You can be prompted for these credentials and save them to a `.netrc` file
```

```python
import earthaccess

auth = earthaccess.login(strategy="netrc")
if not auth:
    auth = earthaccess.login(strategy="interactive", persist=True)
 ```

Once we are authenticated with NASA EDL we can:

* Get a file from a DAAC using a `fsspec` session.

> * Requests temporary S3 credentials from a particular DAAC.

Reviewer: ```suggestion
* Request temporary S3 credentials from a particular DAAC (needed to download or stream data from an S3 bucket in the cloud).
```

* Use the library to download or stream data directly from S3.
* Regenerate CMR tokens (used for restricted datasets)

## Searching for data

Once we have selected our dataset we can search for the data granules using *doi*, *short_name* or *concept_id*.
If are not sure or we don't know ohw to search for a particular dataset we can start with the "searching for data" tutorial or through the data pathfinder program listed above. For a complete list of search parameters we can use visit the API extended documentation.

```python

query = earthaccess.search_data(
    short_name='ATL06',
    version="005",
    cloud_hosted=True,
    bounding_box=(-10, 20, 10, 50),
    temporal=("2020-02", "2020-03")
)

print(f"Granules found: {query.hits()}")

> # We execute our query and ask for 10 results only

Reviewer: ```suggestion
# Execute the query and ask for the first 10 results
```

results = query.get(10)

# If we want all the results we can iterate over items() 
# for results in query.items():
#     print(len(results))

> #     some cool stuff

Reviewer: Is this code staying in? If so, might be helpful for users to clarify the 'cool stuff' line, maybe provide an example?

```

> Now that we have our results we can do multiple things, we can iterate over them to get HTTP (or S3) links; we can download the files to a local folder or we can open these files and stream their content directly to other libraries e.g. xarray.

Reviewer: ```suggestion
Once you have your results you can do multiple things:
1. Iterate over them to get HTTPS (or S3) links.
2. Download the files to a local folder
3. Open these files and stream their content directly to other libraries e.g. xarray. (if working in the cloud)

```

## Accessing the data

> ### Option 1: we only want the links

Reviewer: ```suggestion
### Option 1: Retrieving the HTTPS or S3 links
```

> If we already have a workflow in place for downloading our data, we can use *earthaccess* as a search-only library and get HTTP links from our query results. This could be the case if our current workflow uses a different language and we only need the links as input.

Reviewer: ```suggestion
If you already have a workflow in place for downloading data, you can use *earthaccess* as a search-only library and get the HTTPS links from the query results. This could be useful if your current workflow uses a different language and you only need the links as input.
```

```python

# if the dataset is cloud hosted there will be S3 links available, the access parameter accepts direct or external, direct access is only possible if we are in us-west-2
data_links = [granule.data_links(access="direct") for granule in results]

# or if the data is an on-prem dataset
data_links = [granule.data_links(access="external") for granule in results]

```

> > Note: as a bonus, *earthaccess* can get S3 credentials for us, or auhenticated HTTP sessions in case we want to use them with a different library.

Reviewer: ```suggestion
> Note: as a bonus, *earthaccess* can get S3 credentials for us, or authenticated HTTPS sessions in case we want to use them with a different library.
```

> ### Option 2: we want to download the results to a local folder

Reviewer: ```suggestion
### Option 2: Download the data to a local folder
```

> This option is practical if we have the necessary space available on disk, the library will inform us about the approximate size to be downloaded and its progress.

Reviewer: ```suggestion
This option is practical if you have the necessary space available on disk, the *earthaccess* library will print out the approximate size of the download and its progress.
```

```python
for results in query.items():
    # results is a list of up to 2000 granules (files)
    earthaccess.download(results,
                         local_path="./data/atl06/",
                         auth=auth)

```

> ### Option 3: we want to stream the results directly to xarray

Reviewer: ```suggestion
### Option 3: Direct S3 Access - Stream the data directly to xarray 
```

> This method works better for in-region access when we are working with gridded datasets (processing level 3 and above).

Reviewer: ```suggestion
This method works best if you are in the same region as the data (us-west-2) and you are working with gridded datasets (processing level 3 and above).
```

```python
import xarray as xr

ds = xr.open_mfdataset(earthaccess.open(results, auth=auth), engine="scipy")

```

> And that's it! this code works the same for cloud hosted data and we don't have to change it if we are using DAAC hosted datasets.

Reviewer: Not quite sure what you mean by this sentence. Do you mean you can also use this same command for on-prem data? If so, I suggest the following edit:
'And that's it! Just one line of code, and this same piece of code will also work  for data that are not hosted in the cloud, i.e. on-prem data sets.'

> ## More examples:

Reviewer: Not quite sure what you mean by 'More examples'. Is this list highlighting other capabilities of earthaccess, or examples that you wish to include at a later date in the Readme or readthedocs?
